[WIP] Add `create_unique_workdir` option to the Dask and Local executors

arosen93 opened the pull request:
**Note**: This PR should not be merged or reviewed until #1641 is merged, as this branch is itself a branch of @madhur-tandon's PR.

Addresses #1628 for the Dask and Local executors, mimicking the approach used in the [Slurm executor](https://github.com/AgnostiqHQ/covalent-slurm-plugin/blob/efe2ff943b850901668973c9364e374cd5be3e35/covalent_slurm_plugin/slurm.py#L81) for consistency.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation and CHANGELOG accordingly.
- [x] I have read the CONTRIBUTING document.

madhur-tandon commented on covalent/executor/executor_plugins/dask.py:
```diff
@@ -66,25 +69,42 @@ def __init__(
         conda_env: str = "",
         cache_dir: str = "",
+        workdir: str = "",
+        create_unique_workdir: bool = None,
```
This could be `Optional[bool]` where you do `from typing import Optional` at the top...

madhur-tandon replied:
Alternatively, you could also use the default value as `False`.

arosen93 replied:
I wanted to have it set up such that if it's not specified at all by the user, the default value from the `_EXECUTOR_PLUGIN_DEFAULTS`. If I set it to `False`, then there is no good way to check if that's because the user specified `False` or because the default value was left at `False`. I can change it to `Optional[bool]` since that's the appropriate type here (although in practice, the user probably shouldn't be setting `create_unique_workdir = None` explicitly).

madhur-tandon replied:
Yeah, makes sense, `Optional[bool]` just improves it a bit slightly though 😅

arosen93 replied:
Agreed! Just changed it. Thanks for the note!

madhur-tandon commented on covalent/executor/executor_plugins/local.py:
```diff
@@ -59,6 +63,29 @@ class LocalExecutor(BaseExecutor):
     """
 
+    def __init__(
+        self, workdir: str = "", create_unique_workdir: bool = None, *args, **kwargs
```
Similarly here i.e. `create_unique_workdir: Optional[bool] = None` where you do `from typing import Optional` at the top. Alternatively, you could also use the default value as `False`

madhur-tandon commented:
Hi @arosen93 

Thanks for extending my PR 😄 

I skimmed through and added very minor comments....syntactic sugar basically.
But, when https://github.com/AgnostiqHQ/covalent/pull/1641 gets reviewed and/or merged, you can perhaps rebase on top of it then.

Thanks again!

arosen93 commented:
I kind of screwed up the branch merging here (oops...) so I'll close and start fresh when #1641 is merged :)

madhur-tandon commented:
@arosen93 Let me know if you want me to extend this perhaps?

arosen93 commented:
While I'd certainly by all means appreciate it, you've already done a lot so do whatever you wish! I don't want to burden you with more work. Sadly, I kind of screwed up with git here and I don't think commits aren't visible in this PR anymore :( Luckily, I still had the files locally. I've put the relevant info below both for you and myself.

<details><summary>dask.py</summary>
<p>

```python
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the GNU Affero General Public License 3.0 (the "License").
# A copy of the License may be obtained with this software package or at
#
#      https://www.gnu.org/licenses/agpl-3.0.en.html
#
# Use of this file is prohibited except in compliance with the License. Any
# modifications or derivative works of this file must retain this copyright
# notice, and modified files must contain a notice indicating that they have
# been altered from the originals.
#
# Covalent is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.
#
# Relief from the License may be granted by purchasing a commercial license.

"""
Module for defining a Dask executor that submits the input python function in a dask cluster
and waits for execution to finish then returns the result.

This is a plugin executor module; it is loaded if found and properly structured.
"""

import os
from pathlib import Path
from typing import Callable, Dict, List, Literal, Optional

from dask.distributed import CancelledError, Client, Future

from covalent._shared_files import TaskRuntimeError, logger

# Relative imports are not allowed in executor plugins
from covalent._shared_files.config import get_config, set_config
from covalent._shared_files.exceptions import TaskCancelledError
from covalent._shared_files.utils import _address_client_mapper
from covalent.executor.base import AsyncBaseExecutor
from covalent.executor.utils.wrappers import io_wrapper as dask_wrapper

# The plugin class name must be given by the executor_plugin_name attribute:
EXECUTOR_PLUGIN_NAME = "DaskExecutor"

app_log = logger.app_log
log_stack_info = logger.log_stack_info

_EXECUTOR_PLUGIN_DEFAULTS = {
    "log_stdout": "stdout.log",
    "log_stderr": "stderr.log",
    "cache_dir": os.path.join(
        os.environ.get("XDG_CACHE_HOME") or os.path.join(os.environ["HOME"], ".cache"), "covalent"
    ),
    "workdir": os.path.join(os.environ["HOME"], "covalent", "workdir"),
    "create_unique_workdir": False,
}


class DaskExecutor(AsyncBaseExecutor):
    """
    Dask executor class that submits the input function to a running dask cluster.
    """

    def __init__(
        self,
        scheduler_address: str = "",
        log_stdout: str = "stdout.log",
        log_stderr: str = "stderr.log",
        conda_env: str = "",
        cache_dir: str = "",
        workdir: str = "",
        create_unique_workdir: Optional[bool] = None,
        current_env_on_conda_fail: bool = False,
    ) -> None:
        if not cache_dir:
            cache_dir = _EXECUTOR_PLUGIN_DEFAULTS["cache_dir"]

        if not workdir:
            try:
                workdir = get_config("executors.dask.workdir")
            except KeyError:
                workdir = _EXECUTOR_PLUGIN_DEFAULTS["workdir"]
                info_msg = f"Couldn't find `executors.dask.workdir` in config, creating a default one at {workdir}"
                app_log.info(info_msg)
                print(info_msg)

        if create_unique_workdir is None:
            try:
                create_unique_workdir = get_config("executors.dask.create_unique_workdir")
            except KeyError:
                create_unique_workdir = _EXECUTOR_PLUGIN_DEFAULTS["create_unique_workdir"]

        if not scheduler_address:
            try:
                scheduler_address = get_config("dask.scheduler_address")
            except KeyError:
                app_log.debug(
                    "No dask scheduler address found in config. Address must be set manually."
                )

        super().__init__(log_stdout, log_stderr, conda_env, cache_dir, current_env_on_conda_fail)

        self.scheduler_address = scheduler_address
        self.workdir = workdir
        set_config("executors.dask.workdir", self.workdir)
        self.create_unique_workdir = create_unique_workdir

    async def run(self, function: Callable, args: List, kwargs: Dict, task_metadata: Dict):
        """Submit the function and inputs to the dask cluster"""

        if await self.get_cancel_requested():
            app_log.debug("Task has cancelled")
            raise TaskCancelledError

        dispatch_id = task_metadata["dispatch_id"]
        node_id = task_metadata["node_id"]

        if self.create_unique_workdir:
            current_workdir = os.path.join(self.workdir, dispatch_id, "node_" + str(node_id))
        else:
            current_workdir = self.workdir
        Path(current_workdir).mkdir(parents=True, exist_ok=True)

        dask_client = _address_client_mapper.get(self.scheduler_address)

        if not dask_client:
            dask_client = Client(address=self.scheduler_address, asynchronous=True)
            _address_client_mapper[self.scheduler_address] = dask_client
            await dask_client

        future = dask_client.submit(dask_wrapper, function, args, kwargs, current_workdir)
        await self.set_job_handle(future.key)
        app_log.debug(f"Submitted task {node_id} to dask with key {future.key}")

        try:
            result, worker_stdout, worker_stderr, tb = await future
        except CancelledError:
            raise TaskCancelledError()

        print(worker_stdout, end="", file=self.task_stdout)
        print(worker_stderr, end="", file=self.task_stderr)

        if tb:
            print(tb, end="", file=self.task_stderr)
            raise TaskRuntimeError(tb)

        # FIX: need to get stdout and stderr from dask worker and print them
        return result

    async def cancel(self, task_metadata: Dict, job_handle) -> Literal[True]:
        """
        Cancel the task being executed by the dask executor currently

        Arg(s)
            task_metadata: Metadata associated with the task
            job_handle: Key assigned to the job by Dask

        Return(s)
            True by default
        """
        dask_client = _address_client_mapper.get(self.scheduler_address)

        if not dask_client:
            dask_client = Client(address=self.scheduler_address, asynchronous=True)
            _address_client_mapper[self.scheduler_address] = dask_client
            await dask_client

        fut: Future = Future(key=job_handle, client=dask_client)
        await fut.cancel()
        app_log.debug(f"Cancelled future with key {job_handle}")
        return True
```

</p>
</details>


<details><summary>local.py</summary>
<p>

```python
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the GNU Affero General Public License 3.0 (the "License").
# A copy of the License may be obtained with this software package or at
#
#      https://www.gnu.org/licenses/agpl-3.0.en.html
#
# Use of this file is prohibited except in compliance with the License. Any
# modifications or derivative works of this file must retain this copyright
# notice, and modified files must contain a notice indicating that they have
# been altered from the originals.
#
# Covalent is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.
#
# Relief from the License may be granted by purchasing a commercial license.

"""
Module for defining a local executor that directly invokes the input python function.

This is a plugin executor module; it is loaded if found and properly structured.
"""


import os
from concurrent.futures import ProcessPoolExecutor
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional

# Relative imports are not allowed in executor plugins
from covalent._shared_files import TaskCancelledError, TaskRuntimeError, logger
from covalent._shared_files.config import get_config, set_config
from covalent.executor import BaseExecutor

# Store the wrapper function in an external module to avoid module
# import errors during pickling
from covalent.executor.utils.wrappers import io_wrapper

# The plugin class name must be given by the executor_plugin_name attribute:
EXECUTOR_PLUGIN_NAME = "LocalExecutor"

app_log = logger.app_log
log_stack_info = logger.log_stack_info

_EXECUTOR_PLUGIN_DEFAULTS = {
    "log_stdout": "stdout.log",
    "log_stderr": "stderr.log",
    "cache_dir": os.path.join(
        os.environ.get("XDG_CACHE_HOME") or os.path.join(os.environ["HOME"], ".cache"), "covalent"
    ),
    "workdir": os.path.join(os.environ["HOME"], "covalent", "workdir"),
    "create_unique_workdir": False,
}

proc_pool = ProcessPoolExecutor()


class LocalExecutor(BaseExecutor):
    """
    Local executor class that directly invokes the input function.
    """

    def __init__(
        self, workdir: str = "", create_unique_workdir: Optional[bool] = None, *args, **kwargs
    ) -> None:
        if not workdir:
            try:
                workdir = get_config("executors.local.workdir")
            except KeyError:
                workdir = _EXECUTOR_PLUGIN_DEFAULTS["workdir"]
                info_msg = f"Couldn't find `executors.local.workdir` in config, creating a default one at {workdir}"
                app_log.info(info_msg)
                print(info_msg)

        if create_unique_workdir is None:
            try:
                create_unique_workdir = get_config("executors.dask.create_unique_workdir")
            except KeyError:
                create_unique_workdir = _EXECUTOR_PLUGIN_DEFAULTS["create_unique_workdir"]

        super().__init__(*args, **kwargs)
        self.workdir = workdir
        self.create_unique_workdir = create_unique_workdir
        set_config("executors.local.workdir", self.workdir)

    def run(self, function: Callable, args: List, kwargs: Dict, task_metadata: Dict) -> Any:
        """
        Execute the function locally

        Arg(s)
            function: Function to be executed
            args: Arguments passed to the function
            kwargs: Keyword arguments passed to the function
            task_metadata: Metadata of the task to be executed

        Return(s)
            Task output
        """
        app_log.debug(f"Running function {function} locally")

        self.set_job_handle(42)

        if self.get_cancel_requested():
            app_log.debug("Task has been cancelled don't proceed")
            raise TaskCancelledError

        dispatch_id = task_metadata["dispatch_id"]
        node_id = task_metadata["node_id"]

        if self.create_unique_workdir:
            current_workdir = os.path.join(self.workdir, dispatch_id, "node_" + str(node_id))
        else:
            current_workdir = self.workdir
        Path(current_workdir).mkdir(parents=True, exist_ok=True)

        # Run the target function in a separate process
        fut = proc_pool.submit(io_wrapper, function, args, kwargs, current_workdir)

        output, worker_stdout, worker_stderr, tb = fut.result()

        print(worker_stdout, end="", file=self.task_stdout)
        print(worker_stderr, end="", file=self.task_stderr)

        if tb:
            print(tb, end="", file=self.task_stderr)
            raise TaskRuntimeError(tb)

        return output
```

</p>
</details>

<details><summary>dask test</summary>
<p>

```python
def test_dask_executor_with_unique_workdir(mocker):
    from dask.distributed import LocalCluster

    with tempfile.TemporaryDirectory() as tmp_dir:
        lc = LocalCluster()
        de = ct.executor.DaskExecutor(
            lc.scheduler_address, workdir=tmp_dir, create_unique_workdir=True
        )

        mock_get_cancel_requested = mocker.patch.object(
            de, "get_cancel_requested", AsyncMock(return_value=False)
        )
        mock_set_job_handle = mocker.patch.object(de, "set_job_handle", AsyncMock())

        def simple_task(x, y):
            with open("job.txt", "w") as w:
                w.write(str(x + y))
            return "Done!"

        args = [1, 2]
        kwargs = {}
        task_metadata = {"dispatch_id": "asdf", "node_id": 1}
        result = asyncio.run(de.run(simple_task, args, kwargs, task_metadata))
        w = os.path.join(
            tmp_dir, task_metadata["dispatch_id"], "node_" + str(task_metadata["node_id"])
        )
        assert result == "Done!"
        assert os.listdir(w) == ["job.txt"]
        assert open(os.path.join(w, "job.txt")).read() == "3"

        mock_get_cancel_requested.assert_awaited()
        mock_set_job_handle.assert_awaited()
```

</p>
</details>


<details><summary>local test</summary>
<p>

```python
def test_local_executor_with_unique_workir(mocker):
    with tempfile.TemporaryDirectory() as tmp_dir:
        le = ct.executor.LocalExecutor(workdir=tmp_dir, create_unique_workdir=True)

        mock_set_job_handle = mocker.patch.object(le, "set_job_handle", MagicMock(return_value=42))
        mock_get_cancel_requested = mocker.patch.object(
            le, "get_cancel_requested", MagicMock(return_value=False)
        )

        args = [1, 2]
        kwargs = {}
        task_metadata = {"dispatch_id": "asdf", "node_id": 1}
        assert le.run(simple_task, args, kwargs, task_metadata)
        w = os.path.join(
            tmp_dir, task_metadata["dispatch_id"], "node_" + str(task_metadata["node_id"])
        )
        assert os.listdir(w) == ["job.txt"]
        assert open(os.path.join(w, "job.txt")).read() == "3"

        mock_set_job_handle.assert_called_once()
        mock_get_cancel_requested.assert_called_once()
```

</p>
</details>

madhur-tandon commented:
@kessler-frost / @santoshkumarradha Do you guys recommend to extend the existing PR i.e. https://github.com/AgnostiqHQ/covalent/pull/1641 or perhaps make a new PR just like @arosen93 did with this.

madhur-tandon commented:
I have accommodated changes for create_unique_workdir within https://github.com/AgnostiqHQ/covalent/pull/1641 itself.

CC: @kessler-frost @santoshkumarradha @arosen93

arosen93 commented:
Greatly, greatly appreciated!!!
